### PR TITLE
CB-8040 Parcels are activated without distribution in replaced VM

### PR DESCRIPTION
**Pull request by schfeca75**
Cluster upgrade flow modifications:
1. Determine used parcels:
  - in case of Datalake the only used parcel is CDH
  - in case of Datahub, the used parcels are collected from the blueprint.
2. New step in Cluster upgrade flow: Remove all unused parcels from CM.
3. After cluster runtime upgrade remove all the CDH parcels with different version than the actual one
4. During parcel activation in case of datalake only waiting for the CDH parcel to be activated

### CODE REVIEW

File: core/src/main/java/com/sequenceiq/cloudbreak/service/parcel/ParcelService.java [hunk 1]
```diff
@@ -34,6 +44,33 @@
     @Inject
     private RestClientFactory restClientFactory;
 
+    @Inject
+    private ClusterComponentConfigProvider clusterComponentConfigProvider;
+
+    public Set<ClusterComponent> getParcelComponentsByBlueprint(Stack stack) {
+        Cluster cluster = stack.getCluster();
+        Set<ClusterComponent> components = getParcelComponents(cluster);
+        if (stack.isDatalake()) {
```
Comment (keyki) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/parcel/ParcelService.java:53]: Add logs

File: core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/ClusterUpgradeAvailabilityService.java [hunk 2]
```diff
@@ -169,6 +175,24 @@ private UpgradeV4Response checkForUpgrades(Stack stack, boolean lockComponents)
         return upgradeOptions;
     }
 
+    private Map<String, String> getActivatedParcels(Stack stack) {
+        if (stack.isDatalake()) {
```
Comment (keyki) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/ClusterUpgradeAvailabilityService.java:179]: Add logs

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerModificationService.java [hunk 3]
```diff
@@ -510,15 +512,41 @@ public String getStackRepositoryJson(StackRepoDetails repoDetails, String stackR
...
+        for (ClusterComponent clusterComponent : usedParcelComponents) {
+            ClouderaManagerProduct product = clusterComponent.getAttributes().getSilent(ClouderaManagerProduct.class);
+            cmProducts.put(product.getName(), product);
+        }
+        Map<String, String> installedComponents = gatherInstalledParcels(stack.getName());
+        Set<String> failedDeactivations = new HashSet<>();
+        for (Map.Entry<String, String> installedComp : installedComponents.entrySet()) {
+            if (!cmProducts.containsKey(installedComp.getKey())) {
+                try {
+                    parcelResourceApi.deactivateCommand(stack.getName(), installedComp.getKey(), installedComp.getValue());
```
Comment (keyki) [inline cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerModificationService.java:537]: During upgrade, if DATALAKE
* deactivate ALL except CDH
* undistribute them
* delete them

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/polling/ClouderaManagerPollingServiceProvider.java [hunk 4]
```diff
@@ -106,6 +110,21 @@ public PollingResult startPollingCmParcelActivation(Stack stack, ApiClient apiCl
...
+    public PollingResult startPollingCmParcelStatus(Stack stack, ApiClient apiClient, Map<String, String> parcelVersions,
+            ParcelStatus parcelStatus) {
+        LOGGER.debug("Waiting for Cloudera Manager parcels {} to become to status [{}]. [Server address: {}]", parcelVersions, parcelStatus,
+                stack.getClusterManagerIp());
+        return pollCommandWithTimeListener(stack, apiClient, BigDecimal.ZERO, POLL_FOR_ONE_HOUR,
+                new ClouderaManagerParcelStatusListenerTask(clouderaManagerApiPojoFactory, cloudbreakEventService, parcelVersions, parcelStatus));
+    }
+
+    public PollingResult startPollingCmParcelDelete(Stack stack, ApiClient apiClient, Map<String, String> parcelVersions) {
+        LOGGER.debug("Waiting for Cloudera Manager parcels {} to become to status [{}]. [Server address: {}]", parcelVersions,
```
Comment (keyki) [inline cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/polling/ClouderaManagerPollingServiceProvider.java:122]: 3 param is required, but only 2 is provided

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/polling/task/ClouderaManagerParcelDeletedListenerTask.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    protected boolean doStatusCheck(ClouderaManagerPollerObject pollerObject, CommandsResourceApi commandsResourceApi) throws ApiException {
+        ApiClient apiClient = pollerObject.getApiClient();
+        Stack stack = pollerObject.getStack();
+        ApiParcelList parcels = getClouderaManagerParcels(apiClient, stack.getName());
+        List<ApiParcel> existedParcels = getExistedParcels(parcelVersion, parcels);
+        if (existedParcels.isEmpty()) {
+            LOGGER.debug("Parcels are deleted succesfully.");
+            return true;
+        } else {
+            LOGGER.debug("Some parcels are not yet deleted: [{}].", getJoinedParcelStages(existedParcels));
```
Comment (keyki) [inline cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/polling/task/ClouderaManagerParcelDeletedListenerTask.java:45]: The second upgrade stuck for me at checking the deleted parcels
```
Some parcels are not yet deleted: [(CDH 7.2.1-1.cdh7.2.1.p0.4763198 : AVAILABLE_REMOTELY)].
```
We're checking if it's empty, but for me, it shows AVAILABLE_REMOTELY on the UI. We can consider this one as deleted as well.

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/polling/task/ClouderaManagerParcelDeletedListenerTask.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+            return false;
+        }
+    }
+
+    private ApiParcelList getClouderaManagerParcels(ApiClient apiClient, String stackName) throws ApiException {
+        ParcelsResourceApi parcelsResourceApi = clouderaManagerApiPojoFactory.getParcelsResourceApi(apiClient);
+        return parcelsResourceApi.readParcels(stackName, "summary");
+    }
+
+    private List<ApiParcel> getExistedParcels(Map<String, String> parcelsVersions, ApiParcelList parcels) {
```
Comment (pdarvasi) [inline cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/polling/task/ClouderaManagerParcelDeletedListenerTask.java:55]: typo: existed -> existing

File: core/src/main/java/com/sequenceiq/cloudbreak/service/parcel/ParcelService.java [hunk 7]
```diff
@@ -35,6 +45,35 @@
...
+            ClusterComponent stackComponent = components.stream()
+                    .filter(clusterComponent -> clusterComponent.getName().equals(StackType.CDH.name()))
+                    .findFirst().orElse(null);
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/parcel/ParcelService.java:57]: `.orElseThrow()` would spare us the next line for null check

File: core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/ClusterUpgradeAvailabilityService.java [hunk 8]
```diff
@@ -169,6 +175,26 @@ private UpgradeV4Response checkForUpgrades(Stack stack, boolean lockComponents)
...
+            Set<ClusterComponent> components = clusterComponentConfigProvider.getComponentsByClusterId(stack.getCluster().getId());
+            ClusterComponent stackComponent = components.stream()
+                    .filter(clusterComponent -> clusterComponent.getName().equals(StackType.CDH.name()))
+                    .findFirst().orElse(null);
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/ClusterUpgradeAvailabilityService.java:183]: same here, `orElseThrow`

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerParcelService.java [hunk 9]
```diff
@@ -12,39 +17,146 @@
...
+            }
+        }
+        if (!failedDeactivations.isEmpty()) {
+            throw new ClouderaManagerOperationFailedException(String.format("Deactivation failed on the following products: %s", failedDeactivations));
+        }
+    }
+
+    public void undistributeUnusedParcels(ApiClient apiClient, ParcelsResourceApi parcelsResourceApi, ParcelResourceApi parcelResourceApi, Stack stack,
+            Map<String, ClouderaManagerProduct> cmProducts) {
+        Map<String, String> distributedComponents = getParcelsInStatus(parcelsResourceApi, stack.getName(), ParcelStatus.DISTRIBUTED);
```
[the remark this answers is not in the document]
Comment (schfeca75) [reply cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerParcelService.java:62]: Followup task: https://jira.cloudera.com/browse/CB-8485

**Pull request merged**